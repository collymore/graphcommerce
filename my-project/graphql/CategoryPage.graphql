query CategoryPage($url: String!, $rootCategory: String!) {
  ...MenuQueryFragment
  ...FooterQueryFragment
  ...PageContentQueryFragment

  ...CategoryQueryFragment
}
