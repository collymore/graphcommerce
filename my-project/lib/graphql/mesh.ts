import { getBuiltMesh } from '../../.mesh'

export default getBuiltMesh()
