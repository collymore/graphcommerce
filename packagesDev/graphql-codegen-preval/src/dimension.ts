import { RequestOptions } from 'http'
import { storeToLocale } from '@graphcommerce/magento-store'

type StaticGeneratorProps = {
  operationName: string
  runOperation: (headers: Record<string, string>, variables: Record<string, string>) => Promise<JSON>
  typescriptOperationType: string
  locales: string[]
}

async function generateByStore(props: StaticGeneratorProps) {
  const { operationName, runOperation, typescriptOperationType, locales } = props

  // Create request Record<string, JSON> and run them in parallel
  const requests = locales.map(locale => {
    const headers = {
      'Content-Type': 'application/json',
      'Accept-Language': locale,
    }

    return [locale, runOperation(headers, {})] as const
  })

  // Wait for all requests to finish and map to a Record<string, JSON>

  const responses = await Promise.all(requests);
  

  // Generate function there the locale is passed as a parameter
  return `
    const ${}

    export function ${operationName}(locale: string): ${typescriptOperationType} {
    
  
  `
}
