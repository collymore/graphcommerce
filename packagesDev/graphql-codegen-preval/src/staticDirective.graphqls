type Header {
  key: String!
  value: String!
}

"""
Preval it!
"""
directive @static on QUERY
