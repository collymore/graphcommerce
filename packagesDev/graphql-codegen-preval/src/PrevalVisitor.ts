/* eslint-disable no-underscore-dangle */
/* eslint-disable import/no-extraneous-dependencies */
import { FetchResult } from '@apollo/client'
import { ClientSideBaseVisitor, DocumentMode } from '@graphql-codegen/visitor-plugin-common'
import { pascalCase } from 'change-case-all'
import { OperationDefinitionNode, GraphQLSchema } from 'graphql'

export class PrevalVisitor extends ClientSideBaseVisitor {
  constructor(schema: GraphQLSchema, protected fetchResult: FetchResult) {
    super(schema, [], {}, { documentVariableSuffix: '' }, [])
  }

  OperationDefinition(node: OperationDefinitionNode): string {
    this._collectedOperations.push(node)

    const documentVariableName = this.getOperationVariableName(node)

    const operationType: string = pascalCase(node.operation)
    const operationTypeSuffix: string = this.getOperationSuffix(node, operationType)

    const operationResultType: string = this.convertName(node, {
      suffix: operationTypeSuffix + this._parsedConfig.operationResultSuffix,
    })
    const operationVariablesTypes: string = this.convertName(node, {
      suffix: `${operationTypeSuffix}Variables`,
    })

    let documentString = ''
    if (this.config.documentMode !== DocumentMode.external) {
      // only generate exports for named queries
      if (documentVariableName !== '') {
        documentString = `${
          this.config.noExport ? '' : 'export'
        } const ${documentVariableName}: ${operationResultType} =${
          this.config.pureMagicComment ? ' /*#__PURE__*/' : ''
        } ${JSON.stringify(this.fetchResult.data)};`
      }
    }

    const hasRequiredVariables = this.checkVariablesRequirements(node)

    const additional = this.buildOperation(
      node,
      documentVariableName,
      operationType,
      operationResultType,
      operationVariablesTypes,
      hasRequiredVariables,
    )

    return [documentString, additional].filter((a) => a).join('\n')
  }
}
