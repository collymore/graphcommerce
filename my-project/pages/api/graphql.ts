import { createServer, config } from '@graphcommerce/graphql-mesh'
import mesh from '../../lib/graphql/mesh'

export default await createServer(mesh, '/api/graphql')
export { config }
