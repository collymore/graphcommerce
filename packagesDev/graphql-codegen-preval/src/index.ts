/* eslint-disable no-underscore-dangle */
/* eslint-disable import/no-extraneous-dependencies */
import { FetchResult } from '@apollo/client'
import { oldVisit, PluginFunction, Types } from '@graphql-codegen/plugin-helpers'
import { LoadedFragment, optimizeOperations } from '@graphql-codegen/visitor-plugin-common'
import fetch from 'cross-fetch'
import {
  ASTNode,
  visit,
  BREAK,
  print,
  DocumentNode,
  concatAST,
  FragmentDefinitionNode,
  Kind,
} from 'graphql'
import { PrevalVisitor } from './PrevalVisitor'

export type StaticQueryPluginConfig = { endpoint: string }

const directiveName = 'static'

function isStaticQuery(document: ASTNode) {
  let is = false
  visit(document, {
    Directive: (node) => {
      if (node.name.value === directiveName) is = true
      return BREAK
    },
  })
  return is
}

function stripDirective(document: DocumentNode, name: string) {
  return visit(document, {
    Directive: (node) => (node.name.value === name ? null : undefined),
  })
}

function assertConfig(
  config: Partial<StaticQueryPluginConfig>,
): asserts config is StaticQueryPluginConfig {
  if (!config.endpoint) {
    throw new Error('You need to provide an endpoint for preval')
  }
}

export const plugin: PluginFunction<
  Partial<StaticQueryPluginConfig>,
  Types.ComplexPluginOutput
> = async (schema, documents, config) => {
  assertConfig(config)

  const appends: string[] = []

  for await (const documentFile of documents) {
    const { document } = documentFile
    if (!document) break
    if (!isStaticQuery(document)) break

    const newDocument = stripDirective(document, directiveName)

    const req = await fetch(config.endpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query: print(newDocument),
      }),
    })
    const res = (await req.json()) as FetchResult

    if (res.errors) {
      throw Error(
        `Error while resolving static query: ${res.errors.map((e) => e.message).join(', ')} (${
          documentFile.location
        })`,
      )
    }

    if (res.data) {
      const visitor = new PrevalVisitor(schema, res)
      const visitorResult = oldVisit(newDocument, { leave: visitor }) as { definitions: string[] }

      const results = visitorResult.definitions.filter((t) => typeof t === 'string')

      appends.push(...results)
    }

    documentFile.document = newDocument
  }

  return {
    append: appends,
    content: '',
  }
}
