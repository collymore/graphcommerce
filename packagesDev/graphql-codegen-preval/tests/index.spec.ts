import { buildSchema, parse } from 'graphql'
import { plugin } from '../src'

const testSchema = buildSchema(/* GraphQL */ `
  type Avatar {
    id: ID!
    url: String!
  }

  type User {
    id: ID!
    login: String!
    avatar(height: Int!, width: Int!): Avatar
  }

  type Query {
    user: User!
    users: [User!]!
  }
`)

it('it generates a prevalled query', async () => {
  const testDocument = parse(/* GraphQL */ `
    query StoreConfig @static {
      storeConfig {
        base_url
        absolute_footer
      }
    }
  `)

  const res = await plugin(testSchema, [{ document: testDocument }], {
    endpoint: 'https://backend.reachdigital.dev/graphql',
  })

  expect(res.append?.join('\n')).toMatchSnapshot()
})
