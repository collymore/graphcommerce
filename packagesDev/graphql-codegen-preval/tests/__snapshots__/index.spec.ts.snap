// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates a prevalled query 1`] = `"export const StoreConfig: StoreConfigQuery = {\\"storeConfig\\":{\\"base_url\\":\\"https://backend.reachdigital.dev/\\",\\"absolute_footer\\":null}};"`;

